src/libraries/Native/Unix/System.Native/pal_signal.c

```diff
@@ -17,33 +17,182 @@
 #include <unistd.h>
 
 static pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;
-static struct sigaction g_origSigIntHandler, g_origSigQuitHandler; // saved signal handlers for ctrl handling
-static struct sigaction g_origSigContHandler, g_origSigChldHandler; // saved signal handlers for reinitialization
-static struct sigaction g_origSigWinchHandler; // saved signal handlers for SIGWINCH
-static volatile CtrlCallback g_ctrlCallback = NULL; // Callback invoked for SIGINT/SIGQUIT
-static volatile TerminalInvalidationCallback g_terminalInvalidationCallback = NULL; // Callback invoked for SIGCHLD/SIGCONT/SIGWINCH
-static volatile SigChldCallback g_sigChldCallback = NULL; // Callback invoked for SIGCHLD
+
+// Saved signal handlers
+static struct sigaction* g_origSigHandler;
+static bool* g_handlerIsInstalled;
+
+// Callback invoked for SIGCHLD/SIGCONT/SIGWINCH
+static volatile TerminalInvalidationCallback g_terminalInvalidationCallback = NULL;
+// Callback invoked for SIGCHLD
+static volatile SigChldCallback g_sigChldCallback = NULL;
+static volatile bool g_sigChldConsoleConfigurationDelayed;
+static void (*g_sigChldConsoleConfigurationCallback)(void);
+// Callback invoked for for SIGTTOU while terminal settings are changed.
+static volatile ConsoleSigTtouHandler g_consoleTtouHandler;
+
+// Callback invoked for PosixSignal handling.
+static PosixSignalHandler g_posixSignalHandler = NULL;
+// Tracks whether there are PosixSignal handlers registered.
+static volatile bool* g_hasPosixSignalRegistrations;
+
 static int g_signalPipe[2] = {-1, -1}; // Pipe used between signal handler and worker
 
-static struct sigaction* OrigActionFor(int sig)
+static int GetSignalMax() // Returns the highest usable signal number.
+{
+    return NSIG;
```

Comment:
@wfurt @alexrp do you have a suggestion what value could be used on FreeBSD that would include the real-time signals?

Reply:
The best thing I can think of is using `SIGRTMAX` but it seems there's a bit of a gap of invalid signals between `NSIG` and `SIGRTMIN`...

Reply:
The standard C library may fail `sigaction` for signals that it uses internally, which will surface as an `IOException`.